### DISCUSSION TITLE
Berry display updates and http API for up/downloading files

### DISCUSSION THREAD

**Discussion by msperl** [arendst/Tasmota#15569] [Q&A] [answered]
I am experiencing the following unexpected behaviour with a self-compiled tasmota 11.1.0 for esp-32-lvgl for a lanbon-l8 display:

When there is no berry code running to update the display I can run:
`while true; do curl -s -m 5 -u "$AUTH" 'http://10.10.22.127/ufsd?download=/display.ini' | wc -c ; done`

And always get the length of the display.ini
So:
```
240
240
240
240
240
240
240
240
240
```

But as soon as I run  the following berry code:
```
# start the display
lv.start()

# set background color to blue
scr = lv.scr_act()
scr.set_style_bg_color(lv.color(lv.COLOR_BLUE), lv.PART_MAIN | lv.STATE_DEFAULT)

# create a lv_img object and set it to Tasmota logo
logo = lv.img(scr)
logo.set_tasmota_logo()
logo.center()

# recolor logo to white
logo.set_style_img_recolor_opa(255, lv.PART_MAIN | lv.STATE_DEFAULT)
logo.set_style_img_recolor(lv.color(lv.COLOR_WHITE), lv.PART_MAIN | lv.STATE_DEFAULT)

# rotate by 30 degrees - or 300 deci-degrees
logo.set_angle(300)

cur_zoom = 200
cur_incr = 10
def animate_logo()
  cur_zoom += cur_incr
  if cur_zoom > 600 cur_incr = - cur_incr end
  if cur_zoom < 200 cur_incr = - cur_incr end
  logo.set_zoom(cur_zoom)
  tasmota.set_timer(200, animate_logo)
end
animate_logo()
```
So essentially a logo that is updated(its scale changed) 5 times a second.

I get this behaviour in the curl loop:
```
240
240
240
240
240
240
0
240
0
240
240
240
240
240
240
240
240
240
240
240
240
240
240
240
240
0
240
240
240
240
240
```

If you do a ```curl -vvv``` to get all the details it looks like this:
```
*   Trying 10.10.22.127:80...
* Connected to 10.10.22.127 (10.10.22.127) port 80 (#0)
* Server auth using Basic with user 'admin'
> GET /ufsd?download=/display.ini HTTP/1.1
> Host: 10.10.22.127
> Authorization: Basic YWRt…
> User-Agent: curl/7.74.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Content-Type: application/octet-stream
< Content-Disposition: attachment; filename=display.ini
< Content-Length: 240
< Connection: close
<
:H,ST7789,240,320,16,SPI,1,22,19,23,21,-1,18,25,40
:S,2,1,3,0,80,30
:I
01,A0
11,A0
3A,81,55
36,81,00
21,80
13,80
29,A0
:o,28
:O,29
:A,2A,2B,2C
:B,60,1
:R,36
:0,C0,00,00,00
:1,A0,00,00,01
:2,00,00,00,02
:3,60,00,00,03
:i,20,21
:TI1,38,0,4
#
* Closing connection 0
*   Trying 10.10.22.127:80...
* Connected to 10.10.22.127 (10.10.22.127) port 80 (#0)
* Server auth using Basic with user 'admin'
> GET /ufsd?download=/display.ini HTTP/1.1
> Host: 10.10.22.127
> Authorization: Basic YWRt…
> User-Agent: curl/7.74.0
> Accept: */*
>
* Empty reply from server
* Connection #0 to host 10.10.22.127 left intact
*   Trying 10.10.22.127:80...
* Connected to 10.10.22.127 (10.10.22.127) port 80 (#0)
* Server auth using Basic with user 'admin'
> GET /ufsd?download=/display.ini HTTP/1.1
> Host: 10.10.22.127
> Authorization: Basic YWRt…
> User-Agent: curl/7.74.0
> Accept: */*
>
* Empty reply from server
* Connection #0 to host 10.10.22.127 left intact
*   Trying 10.10.22.127:80...
* Connected to 10.10.22.127 (10.10.22.127) port 80 (#0)
* Server auth using Basic with user 'admin'
> GET /ufsd?download=/display.ini HTTP/1.1
> Host: 10.10.22.127
> Authorization: Basic YWRt…
> User-Agent: curl/7.74.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Content-Type: application/octet-stream
< Content-Disposition: attachment; filename=display.ini
< Content-Length: 240
< Connection: close
<
:H,ST7789,240,320,16,SPI,1,22,19,23,21,-1,18,25,40
:S,2,1,3,0,80,30
:I
01,A0
11,A0
3A,81,55
36,81,00
21,80
13,80
29,A0
:o,28
:O,29
:A,2A,2B,2C
:B,60,1
:R,36
:0,C0,00,00,00
:1,A0,00,00,01
:2,00,00,00,02
:3,60,00,00,03
:i,20,21
:TI1,38,0,4
#
* Closing connection 0
```

See that sometimes there is an `* Empty reply from server`

I guess/fear I have also seen the same thing on the normal HTTP-API used for Configuration, but that I can not confirm just as easily...

To me it seems as if the Berry timer handler is blocking/aborting the http requests sometimes.

```Weblog 4``` does not show anything out of the ordinary either besides lines of the following type:
```
13:39:40.575 HTP: Manage File system
13:39:40.691 HTP: Manage File system
13:39:40.809 HTP: Manage File system
13:39:40.942 HTP: Manage File system
13:39:41.032 BRY: GC from 10316 to 5199 bytes, objects freed 66/57 (in 2 ms) - slots from 180/229 to 50/167
13:39:41.108 HTP: Manage File system
13:39:41.228 HTP: Manage File system
13:39:41.383 HTP: Manage File system
13:39:41.489 HTP: Manage File system
13:39:41.606 HTP: Manage File system
```

Is this "normal/expected" behavior?
Has someone else also seen this?
Is there something we can do about it to improve the situation?

Say: Would it be possible to dedicate one of the 2 cores of the ESP32 to Berry or something similar? (just a random thought without knowledge of how the 2 Cores really play along in Tasmota32)

Some background: I came across this when I started implementing some file management for tasmota in ansible using the API...

**Comment by msperl** [comment]
another observtion: in the case things fail (so no content is returned) curl returns with an exit code of 52, with the man page defines as:
```
       52     The server didn't reply anything, which here is considered an error.
```

**Reply by msperl** [reply]
unfortunately the same behavior occurs when there is no file with the given name - no http error is returned, but the http connection is aborted!
```
curl -s -m 5 -u "$AUTH" 'http://10.10.22.127/ufsd?download=/no_such_file'; echo $?
52
```
so there is no means to distinguish between the 2 situations, but as said: I fear that this happens on all HTTP-APIs, and I can rreliably reproduce it only here...

**Comment by s-hadinger** [comment, marked as answer]
The problem is not so much in Berry, but more in LVGL that can consume quite some CPU when refreshing the display: it blocks the entire Tasmota hence the HTTP server.

Unfortunetaly, neither LVGL nor Berry nor Arduino framework nor Tasmota are thread safe. For simplicity they all run in the same thread (RTOS task). I tried to run LVGL in a seperate task with locks between LVGL and Berry, and it creates more problems than it solves, like displaying a screen while configuring the elements in Berry is not yet complete. It may not be impossible, but would require a **LOT** of work.

I would love to have the HTTP server running in another task, but it would not solve your problem, since in-fine the code would run in the Tasmota task.

Sorry but I don't have any easy solution. Tasmota was designed for MQTT in mind, using MQTT will work just fine.

**Reply by msperl** [reply]
I understand and thanks for that feedback.
Management of files via MQTT is not possible - or so it seems, so I can not do those things this way...

As a workarround: Is there maybe some way to stop lvgl or berry for some time with a command to avoid this to happen when I run the script maintaining the files on a tasmota device? 

Maybe something similar to Berry: ```display.started()``` say: ```display.sleep([sleep_ms:int])```
and that would take lvgl updates out of the loop for a well defined period of time.

I also see/understand the "big" picture with concurrency and I guess that having the ability to run different threads in a safe manner would be the best solution, but that is hard work...

**Reply by s-hadinger** [reply]
I'm ok to add `display.pause(true/false)` to stop updating LVGL. The timer will need to be managed in berry like:

``` berry
display.pause(true)    # pause
tasmota.set_timer(2000, /-> display.pause(false))    # unpause in 2 seconds
```

**Reply by sfromis** [reply]
What I might try would be to drop the usage of curl, to instead (via MQTT) trigger a small Berry program to fetch a file from a server. Something like this:
```
def downloadfile(urlpath, filename)
  var wc = webclient()
  wc.begin(urlpath + filename)
  var status = wc.GET()
  if status != 200 
    raise 'connection_error', 'return code: '..str(status) 
  end
  wc.write_file(filename)
  wc.close()
end
downloadfile('http://my-server.lan/berry/', 'myscript.be')
```
Not sure how it would interact with LVGL, but at least the Tasmota side has control over when to initiate communication.

**Reply by s-hadinger** [reply]
This will be more reliable for sure

**Reply by sfromis** [reply]
If the file is not too big, you could even have all the data in the command published via MQTT, to invoke a small program to write a parameter to a file.

**Reply by s-hadinger** [reply]
You don't need `display.pause()` anymore?

**Reply by msperl** [reply]
I would need to rework everything I have done so far to provision now via mqtt - and I am really provisioning including Firmware upgrade, doing a full reset 4 when a template is not right and such .. (and that looses mqtt info)...

I would still be running into issues on those deployment paths.

So I guess that pausing lvgl is still something that would help me avoid those threading shortcomings - if it is not too much work then it would be appreciated if you could implement it! Thanks

**Reply by msperl** [reply]
As for  running an initial command in berry to download something via http: how would I trigger a berry command/script via the command interface?

All I figured out is: 'http://10.10.22.127/bc?c2=2&c1=lv.disp().get_inactive_time()'

But to bootstrap a host for loading berry scripts or similar with mqtt I would need to be able to run something with 'http://10.10.22.127/cs?c2=3&c1=berry "lv.disp().get_inactive_time()"'

Probably I must be missing something.

**Reply by sfromis** [reply]
Like any other command, you can use the `Br` command to execute some Berry code. Do not try to emulate the internal calls done by the console, but use the documented 
`http://10.10.22.127/cm?cmnd=Br+lv.disp().get_inactive_time()` (or whatever)
You could also trigger the file copy this way, MQTT is not strictly required.

**Reply by s-hadinger** [reply]
+1

Please use `Br <cmd>`, what you used is not supported. 

This fits nicely with MQTT

**Reply by msperl** [reply]
thanks - I just did not find that in the documentation - at least not where I was searching

**Reply by sfromis** [reply]
https://tasmota.github.io/docs/Commands/#with-web-requests

**Reply by msperl** [reply]
that one I know.
It is the reference to `BR`that I could not find

**Reply by sfromis** [reply]
While it is not much, the `Br` command is used in the Berry Cookbook page.
